src/coreComponents/linearAlgebra/interfaces/trilinos/TpetraMatrix.cpp

```diff
@@ -0,0 +1,795 @@
+                        globalIndex const * colIndices,
+                        real64 const * values,
+                        localIndex const numRows,
+                        localIndex const numCols )
+{
+  for( localIndex i = 0; i < numRows; ++i )
+  {
+    set( rowIndices[i], colIndices, values + i * numCols, numCols );
+  }
+}
+
+void TpetraMatrix::insert( globalIndex const * rowIndices,
+                           globalIndex const * colIndices,
+                           real64 const * values,
+                           localIndex const numRows,
+                           localIndex const numCols )
+{
+  for( localIndex i = 0; i < numRows; ++i )
+  {
+    insert( rowIndices[i], colIndices, values + i * numCols, numCols );
+  }
+}
+
+void TpetraMatrix::apply( TpetraVector const & src,
+                          TpetraVector & dst ) const
+{
+  GEOSX_LAI_ASSERT( ready() );
+  GEOSX_LAI_ASSERT( src.ready() );
+  GEOSX_LAI_ASSERT( dst.ready() );
+  GEOSX_LAI_ASSERT_EQ( numGlobalRows(), dst.globalSize() );
+  GEOSX_LAI_ASSERT_EQ( numGlobalCols(), src.globalSize() );
+
+  m_matrix->apply( src.unwrapped(),
+                   dst.unwrapped(),
+                   Teuchos::ETransp::NO_TRANS,
+                   1.0,
+                   0.0 );
+}
+
+void TpetraMatrix::applyTranspose( TpetraVector const & src,
+                                   TpetraVector & dst ) const
+{
+  GEOSX_LAI_ASSERT( ready() );
+  GEOSX_LAI_ASSERT( src.ready() );
+  GEOSX_LAI_ASSERT( dst.ready() );
+  GEOSX_LAI_ASSERT_EQ( numGlobalCols(), dst.globalSize() );
+  GEOSX_LAI_ASSERT_EQ( numGlobalRows(), src.globalSize() );
+
+  m_matrix->apply( src.unwrapped(),
+                   dst.unwrapped(),
+                   Teuchos::ETransp::TRANS,
+                   1.0,
+                   0.0 );
+}
+
+void TpetraMatrix::multiply( TpetraMatrix const & src,
+                             TpetraMatrix & dst ) const
+{
+  this->multiply( false, src, false, dst );
+}
+
+void TpetraMatrix::leftMultiplyTranspose( TpetraMatrix const & src,
+                                          TpetraMatrix & dst ) const
+{
+  this->multiply( true, src, false, dst );
+}
+
+void TpetraMatrix::rightMultiplyTranspose( TpetraMatrix const & src,
+                                           TpetraMatrix & dst ) const
+{
+  src.multiply( false, *this, true, dst );
+}
+
+void TpetraMatrix::create( Tpetra_CrsMatrix const & src )
+{
+  GEOSX_LAI_ASSERT( closed() );
+  reset();
+
+  m_matrix = std::make_unique< Tpetra_CrsMatrix >( src );
+  m_src_map = std::make_unique< Tpetra_Map >( *m_matrix->getDomainMap() );
+  m_dst_map = std::make_unique< Tpetra_Map >( *m_matrix->getRangeMap() );
+  m_assembled = true;
+}
+
+void TpetraMatrix::multiplyRAP( TpetraMatrix const & R,
+                                TpetraMatrix const & P,
+                                TpetraMatrix & dst ) const
+{
+  // TODO: Tpetra::TripleMatrixMultiply::MultiplyRAP gives wrong results, investigate
```

Comment:
Strange ... does it work for the symmetric case PtAP?

Reply:
It probably does since I left that one unchanged. I forgot what exactly the problem was here, I think something related to column map of the result being incorrect. I haven't reported it to Trilinos yet because I need to invest some time in constructing a minimal reproducer.